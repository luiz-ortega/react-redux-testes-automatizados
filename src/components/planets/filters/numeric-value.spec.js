import React from 'react';
import { screen, render, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import FilterByNumericValue from './numeric-value';

const submitComparisonInfoMock = jest.fn();

function renderComponent(overrides = {}) {
  return render(
    <FilterByNumericValue submitComparisonInfo={submitComparisonInfoMock} />
  );
}

const select1 = [
  'Surface Water',
  'Diameter',
  'Orbital Period',
  'Rotation Period',
  'Population',
];

const select2 = ['Greater than', 'Less than', 'Equals'];

describe('Component: FilterByNumericValue', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('Displays all elements', () => {
    it('should match snapshot', () => {
      const { baseElement } = renderComponent();
      expect(baseElement).toMatchSnapshot();
    });
  });

  describe(`Component's behavior`, () => {
    it('should NOT submit any information if value is not provided', async () => {
      renderComponent();

      await userEvent.click(screen.getByRole('button', { name: /apply/i }));

      expect(submitComparisonInfoMock).toHaveBeenCalledTimes(0);
    });

    it('should submit user provided information and the initial selected value in the select fields', async () => {
      renderComponent();

      const inputField = screen.getByRole('spinbutton');
      await fireEvent.change(inputField, { target: { value: 22 } });
      await userEvent.click(screen.getByRole('button', { name: /apply/i }));

      expect(submitComparisonInfoMock).toHaveBeenCalledTimes(1);
      expect(submitComparisonInfoMock).toHaveBeenCalledWith(
        'surface_water',
        'greater_than',
        '22'
      );
    });

    it('should submit user provided information on all fields fields', async () => {
      renderComponent();

      const [propertySelector, comparisonSelector] = screen.getAllByRole(
        'combobox'
      );
      const inputField = screen.getByRole('spinbutton');

      await userEvent.selectOptions(propertySelector, 'diameter');
      await userEvent.selectOptions(comparisonSelector, 'less_than');
      await fireEvent.change(inputField, { target: { value: 2000 } });
      await userEvent.click(screen.getByRole('button', { name: /apply/i }));

      expect(submitComparisonInfoMock).toHaveBeenCalledTimes(1);
      expect(submitComparisonInfoMock).toHaveBeenCalledWith(
        'diameter',
        'less_than',
        '2000'
      );
    });

    it('should return values to their original after submitting', async () => {
      renderComponent();

      const [propertySelector, comparisonSelector] = screen.getAllByRole(
        'combobox'
      );
      const inputField = screen.getByRole('spinbutton');

      await userEvent.selectOptions(propertySelector, 'diameter');
      await userEvent.selectOptions(comparisonSelector, 'less_than');
      await fireEvent.change(inputField, { target: { value: 2000 } });
      await userEvent.click(screen.getByRole('button', { name: /apply/i }));

      expect(inputField.value).toEqual('0');
      expect(propertySelector.value).toEqual('surface_water');
      expect(comparisonSelector.value).toEqual('greater_than');
    });
  });
});
