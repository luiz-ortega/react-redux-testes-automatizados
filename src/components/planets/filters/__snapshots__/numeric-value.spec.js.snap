// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: FilterByNumericValue Displays all elements should match snapshot 1`] = `
<body>
  <div>
    <form
      class="px-4 py-5 sm:p-6 text-gray-900"
      data-testid="filter-numeric-value-form"
    >
      <p
        class="block text-sm font-bold uppercase ml-1 text-gray-700"
      >
        Comparison filter
      </p>
      <div
        class="grid grid-flow-row md:grid-flow-col gap-x-2 mb-2"
      >
        <select
          class="capitalize mt-1 block w-full pl-3 pr-10 py-2 text-base border-gray-300 focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm rounded-md"
          data-testid="column-filter"
          id="Columns"
          name="Columns"
        >
          <option
            value="surface_water"
          >
            Surface Water
          </option>
          <option
            value="diameter"
          >
            Diameter
          </option>
          <option
            value="orbital_period"
          >
            Orbital Period
          </option>
          <option
            value="rotation_period"
          >
            Rotation Period
          </option>
          <option
            value="population"
          >
            Population
          </option>
        </select>
        <select
          class="capitalize mt-1 block w-full pl-3 pr-10 py-2 text-base border-gray-300 focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm rounded-md"
          data-testid="comparison-filter"
          id="Comparison"
          name="Comparison"
        >
          <option
            value="greater_than"
          >
            Greater than
          </option>
          <option
            value="less_than"
          >
            Less than
          </option>
          <option
            value="equals"
          >
            Equals
          </option>
        </select>
      </div>
      <input
        autocomplete="off"
        class="shadow-sm focus:ring-indigo-500 focus:border-indigo-500 block w-full sm:text-sm border-gray-300 rounded-md"
        data-testid="value-filter"
        id="value"
        name="value"
        placeholder="Type the information to compare"
        type="number"
        value="0"
      />
      <button
        class="mt-2 inline-flex items-center px-3 py-2 border border-gray-300 shadow-sm text-sm leading-4 font-medium rounded-md text-gray-700 bg-white hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
        data-testid="button-filter"
        type="submit"
      >
        Apply
      </button>
    </form>
  </div>
</body>
`;
